lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -98,181 +128,459 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                    memoryForUpdateTask,
+                    coresPerMapper,
+                    coresForUpdateTask);
             Logger.Log(Level.Info, msg);
         }
 
+        #region IDriverStarted
         /// <summary>
-        /// Requests for evaluator for update task
+        /// Requests evaluators when driver starts
         /// </summary>
         /// <param name="value">Event fired when driver started</param>
         public void OnNext(IDriverStarted value)
         {
-            RequestUpdateEvaluator();
-            //// TODO[REEF-598]: Set a timeout for this request to be satisfied. If it is not within that time, exit the Driver.
+            _evaluatorManager.RequestUpdateEvaluator();
+            _evaluatorManager.RequestMapEvaluators(_totalMappers);
         }
+        #endregion IDriverStarted
 
+        #region IAllocatedEvaluator
         /// <summary>
-        /// Specifies context and service configuration for evaluator depending
-        /// on whether it is for Update function or for map function
+        /// IAllocatedEvaluator handler. It will take the following action based on the system state:
+        /// Case WaitingForEvaluator
+        ///    Add Evaluator to the Evaluator Manager
+        ///    submit Context and Services
+        /// Case Fail
+        ///    Do nothing. This is because the code that sets system Fail has executed FailedAction. It has shut down all the allocated evaluators/contexts. 
+        ///    If a new IAllocatedEvaluator comes after it, we should not submit anything so that the evaluator is returned.
+        /// Other cases - not expected
         /// </summary>
         /// <param name="allocatedEvaluator">The allocated evaluator</param>
         public void OnNext(IAllocatedEvaluator allocatedEvaluator)
         {
-            var configs =
-                _serviceAndContextConfigurationProvider.GetContextConfigurationForEvaluatorById(allocatedEvaluator.Id);
-            allocatedEvaluator.SubmitContextAndService(configs.Context, configs.Service);
+            Logger.Log(Level.Verbose, "AllocatedEvaluator EvaluatorBatchId [{0}], memory [{1}], systemState {2}.", allocatedEvaluator.EvaluatorBatchId, allocatedEvaluator.GetEvaluatorDescriptor().Memory, _systemState.CurrentState);
+            lock (_lock)
+            {
+                switch (_systemState.CurrentState)
+                {
+                    case SystemState.WaitingForEvaluator:
+                        _evaluatorManager.AddAllocatedEvaluator(allocatedEvaluator);
+                        SubmitContextAndService(allocatedEvaluator);
+                        break;
+                    case SystemState.Fail:
+                        Logger.Log(Level.Info, "Receiving IAllocatedEvaluator event, but system is in FAIL state, ignore it.");
+                        break;
+                    default:
+                        UnexpectedState(allocatedEvaluator.Id, "IAllocatedEvaluator");
```

Comment:
Wouldn't it be easier to just throw an exception here? That way, we fail the Driver and get a full exception stack.